pgjdbc/src/main/java/org/postgresql/xa/PGXAConnection.java

```diff
@@ -584,6 +584,18 @@ public boolean setTransactionTimeout(int seconds) {
     return false;
   }
 
+  private int mapSQLStateToXAErrorCode(SQLException sqlException) {
+    if (isPostgreSQLIntegrityConstraintViolation(sqlException)) {
+      return XAException.XA_RBROLLBACK;
+    }
+
+    return XAException.XAER_RMFAIL;
+  }
+
+  private boolean isPostgreSQLIntegrityConstraintViolation(SQLException sqlException) {
+    return sqlException instanceof PSQLException && sqlException.getSQLState().matches("23\\d{3}");
```

Comment:
While `regexp` looks clever, it is actually specified to mean that? What if the code clashes somehow?

Reply:
Correct. Replaced it with a non-regex version.